amp/sharer.go

```diff
@@ -0,0 +1,159 @@
+// seed.
+func SeedSharerFromRoot(root *Share) *SeedSharer {
+	return initSeedSharer(root, root)
+}
+
+func initSeedSharer(root, curr *Share) *SeedSharer {
+	return &SeedSharer{
+		root: *root,
+		curr: *curr,
+	}
+}
+
+// Seed returns the sharer's seed, the primary source of entropy for deriving
+// shares of the root.
+func (s *SeedSharer) Root() Share {
+	return s.root
+}
+
+// Split constructs two child Sharers whose shares sum to the parent Sharer.
+// This allows an HTLC whose payment amount could not be routed to be
+// recursively split into smaller subpayments. After splitting a sharer the
+// parent share should no longer be used, and the caller should use the Child
+// method on each to derive preimage/hash pairs for the HTLCs.
+func (s *SeedSharer) Split() (Sharer, Sharer, error) {
+	shareLeft, shareRight, err := split(&s.curr)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	left := initSeedSharer(&s.root, &shareLeft)
+	right := initSeedSharer(&s.root, &shareRight)
+
+	return left, right, nil
+}
+
+// Child derives a preimage/hash pair to be used for an AMP HTLC.
+// All children of s will use the same underlying share, but have unique unique
+// preimage and hash. This can be used to rerandomize the preimage/hash pair for
+// a given HTLC if a new route is needed.
+func (s *SeedSharer) Child(index uint32) *Child {
+	desc := ChildDesc{
+		Share: s.curr,
+		Index: index,
+	}
+
+	return DeriveChild(s.root, desc)
+}
+
+// ReconstructChildren derives the set of children hashes and preimages from the
+// provided descriptors. The shares from each child descriptor are first used to
+// compute the root, afterwards the child hashes and preimages are
+// determinstically computed. For child descriptor at index i in the input, it's
+// derived child will occupy index i of the returned children.
+func ReconstructChildren(descs ...ChildDesc) []*Child {
+	// Recompute the root by XORing the provided shares.
+	var root Share
+	for _, desc := range descs {
+		xor(&root, &root, &desc.Share)
+	}
+
+	// With the root computed, derive the child hashes and preimages from
+	// the child descriptors.
+	children := make([]*Child, len(descs))
+	for i, desc := range descs {
+		children[i] = DeriveChild(root, desc)
+	}
+
+	return children
+}
+
+// split splits a share into two random values, that when XOR'd reproduce the
+// original share. Given a share s, the two shares are dervied as:
+//   left <-$- random
+//   right = parent ^ left.
+//
+// When reconstructed, we have that:
+//   left ^ right = left ^ parent ^ left
+//                = parent.
+func split(parent *Share) (Share, Share, error) {
+	// Generate a random share for the left child.
+	var left Share
+	if _, err := rand.Read(left[:]); err != nil {
+		return Share{}, Share{}, err
+	}
+
+	// Compute right = parent ^ left.
+	var right Share
+	xor(&right, parent, &left)
+
+	return left, right, nil
+}
+
+// xor stores the byte-wise xor of shares x and y in z.
+func xor(z, x, y *Share) {
```

Comment:
Would it be worth adding this as a method on the `Share` type (to be the receiver of the result `z`)? The syntax would then more resemble the point operations in `btcec`. But maybe that syntax is what we want to avoid in the first place?

Reply:
sgtm, no specific it was does this way originally :) done